MobileCenter/MobileCenter/Internals/Storage/MSLogDBStorage.m

```diff
@@ -0,0 +1,253 @@
+  }
+  return self;
+}
+
+- (instancetype)initWithCapacity:(NSUInteger)capacity {
+  if ((self = [self init])) {
+    _capacity = capacity;
+  }
+  return self;
+}
+
+#pragma mark - Save logs
+
+- (BOOL)saveLog:(id<MSLog>)log withGroupId:(NSString *)groupId {
+  if (!log) {
+    return NO;
+  }
+
+  // Insert this log to the DB.
+  NSData *logData = [NSKeyedArchiver archivedDataWithRootObject:log];
+  NSString *base64Data = [logData base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithLineFeed];
+  NSString *addLogQuery =
+      [NSString stringWithFormat:@"INSERT INTO \"%@\" (\"%@\", \"%@\") VALUES ('%@', '%@')", kMSLogTableName,
+                                 kMSGroupIdColumnName, kMSLogColumnName, groupId, base64Data];
+  BOOL succeeded = [self executeNonSelectionQuery:addLogQuery];
+  NSUInteger logCount = [self countLogsWithGroupId:groupId];
+
+  // Max out DB.
+  if (succeeded && logCount > self.capacity) {
+    NSUInteger overflowCount = logCount - self.capacity;
+    [self deleteOldestLogsWithGroupId:groupId count:overflowCount];
+    MSLogDebug([MSMobileCenter logTag], @"Log storage was over capacity, %ld oldest log(s) deleted.",
+               (long)overflowCount);
+  }
+  return succeeded;
+}
+
+#pragma mark - Load logs
+
+- (BOOL)loadLogsWithGroupId:(NSString *)groupId
+                      limit:(NSUInteger)limit
+             withCompletion:(nullable MSLoadDataCompletionBlock)completion {
+  BOOL logsAvailable = NO;
+  BOOL moreLogsAvailable = NO;
+  NSString *batchId;
+  NSMutableArray<NSArray *> *logEntries;
+  NSMutableArray<NSNumber *> *dbIds = [NSMutableArray<NSNumber *> new];
+  NSMutableArray<id<MSLog>> *logs = [NSMutableArray<id<MSLog>> new];
+
+  // Get ids from batches.
+  NSMutableArray<NSNumber *> *idsInBatches = [NSMutableArray<NSNumber *> new];
+  for (NSString *batchKey in [self.batches allKeys]) {
+    if ([batchKey hasPrefix:groupId]) {
+      [idsInBatches addObjectsFromArray:(NSArray<NSNumber *> * _Nonnull)self.batches[batchKey]];
+    }
+  }
+
+  // Build the "WHERE" clause's condition.
+  NSMutableString *condition = [NSMutableString stringWithFormat:@"\"%@\" = '%@'", kMSGroupIdColumnName, groupId];
+
+  // Take only logs that are not already part of a batch.
+  if (idsInBatches.count > 0) {
+    [condition appendFormat:@" AND \"%@\" NOT IN (%@)", kMSIdColumnName, [idsInBatches componentsJoinedByString:@", "]];
+  }
+
+  /*
+   * There is a need to determine if there will be more logs available than those under the limit.
+   * This is just about knowing if there is at least 1 log above the limit.
+   *
+   * FIXME: We should simply use a count API from the consumer object instead of the "limit + 1" technique, it only
+   * requires 1 SQL request instead of 2 for the count but it is a bit confusing and doesn't realy fit a database
+   * storage.
+   */
+  [condition appendFormat:@" LIMIT %lu", (unsigned long)((limit < NSUIntegerMax) ? limit + 1 : limit)];
```

Comment:
I have no strong opinion on this. 
I think having a 2nd query would make it more "DB usage"-like, but it's more efficient without it. I tend towards making a 2nd query, but I'm open to your opinions.